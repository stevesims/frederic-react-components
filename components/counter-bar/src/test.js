import React from 'react';
import { HashRouter, Link } from 'react-router-dom';
import { mount } from 'enzyme';
import { createMatchers } from 'jest-emotion';
import * as emotion from 'emotion';
import { LINK_COLOUR, WHITE } from 'govuk-colours';

import CounterBar from '.';

expect.extend(createMatchers(emotion));

describe('CounterBar', () => {
  let wrapper;

  it('renders counters without an active total', () => {
    wrapper = mount(<CounterBar>
      <CounterBar.Total score={1}>All counters</CounterBar.Total>
      <CounterBar.Counters>
        <CounterBar.Counter score={1}>Counter 1</CounterBar.Counter>
        <CounterBar.Counter />
      </CounterBar.Counters>
    </CounterBar>);
    const totalWrapper = wrapper.find('Total');
    expect(totalWrapper).not.toHaveStyleRule('background', LINK_COLOUR);
    expect(totalWrapper).not.toHaveStyleRule('color', WHITE);
    expect(totalWrapper).not.toHaveStyleRule('outline', `2px solid ${LINK_COLOUR}`);
  });

  it('renders empty counters', () => {
    expect(wrapper.find('Counter').at(1).exists()).toBe(true);
  });
    
  it('renders counters with an active total', () => {
    wrapper = mount(<CounterBar>
      <CounterBar.Total score={1} active>All counters</CounterBar.Total>
      <CounterBar.Counters>
        <CounterBar.Counter score={1}>Counter 1</CounterBar.Counter>
      </CounterBar.Counters>
    </CounterBar>);
    const totalWrapper = wrapper.find('Total');
    expect(totalWrapper).toHaveStyleRule('background', LINK_COLOUR);
    expect(totalWrapper).toHaveStyleRule('color', WHITE);
    expect(totalWrapper).toHaveStyleRule('outline', `2px solid ${LINK_COLOUR}`);
  });

  it('renders counters with disabled counter values', () => {
    wrapper = mount(<CounterBar>
      <CounterBar.Total active score={0}>All counters</CounterBar.Total>
      <CounterBar.Counters>
        <CounterBar.Counter score={0}>Counter 1</CounterBar.Counter>
      </CounterBar.Counters>
    </CounterBar>);
    const counterWrapper = wrapper.find('CounterWrapper').first();
    expect(counterWrapper.prop('disabled')).toBe(true);
  });
  
  it('renders counters with active values', () => {
    wrapper = mount(<CounterBar>
      <CounterBar.Total score={1}>All counters</CounterBar.Total>
      <CounterBar.Counters>
        <CounterBar.Counter score={1} active>Counter 1</CounterBar.Counter>
      </CounterBar.Counters>
    </CounterBar>);
    const counterWrapper = wrapper.find('Counter').first();
    expect(counterWrapper).toHaveStyleRule('background', LINK_COLOUR);
    expect(counterWrapper).toHaveStyleRule('color', WHITE);
    expect(counterWrapper).toHaveStyleRule('outline', `2px solid ${LINK_COLOUR}`);
  });

  it('renders custom colours on totals', () => {
    wrapper = mount(<CounterBar>
      <CounterBar.Total score={1} scoreColor="yellow" scoreBackgroundColor="pink">All counters</CounterBar.Total>
      <CounterBar.Counters>
        <CounterBar.Counter score={1}>Counter 1</CounterBar.Counter>
      </CounterBar.Counters>
    </CounterBar>);
    const counterWrapper = wrapper.find('ResultCount').first();
    expect(counterWrapper).toHaveStyleRule('color', 'yellow');
    expect(counterWrapper).toHaveStyleRule('background', 'pink');
  });

  it('renders custom colours on counters', () => {
    wrapper = mount(<CounterBar>
      <CounterBar.Total score={1}>All counters</CounterBar.Total>
      <CounterBar.Counters>
        <CounterBar.Counter score={1} scoreColor="orange" scoreBackgroundColor="blue">Counter 1</CounterBar.Counter>
      </CounterBar.Counters>
    </CounterBar>);
    const counterWrapper = wrapper.find('ResultCount').at(1);
    expect(counterWrapper).toHaveStyleRule('background', 'blue');
    expect(counterWrapper).toHaveStyleRule('color', 'orange');
  });

  it('accepts an HTML element string for the total', () => {
    wrapper = mount(<CounterBar>
      <CounterBar.Total score={15} component="aside">All counters</CounterBar.Total>
      <CounterBar.Counters>
        <CounterBar.Counter score={1}>Counter 1</CounterBar.Counter>
        <CounterBar.Counter score={2}>Counter 2</CounterBar.Counter>
        <CounterBar.Counter score={3}>Counter 3</CounterBar.Counter>
        <CounterBar.Counter score={4}>Counter 4</CounterBar.Counter>
        <CounterBar.Counter score={5}>Counter 5</CounterBar.Counter>
      </CounterBar.Counters>
    </CounterBar>);
    const totalWrapper = wrapper.find('Total');
    expect(totalWrapper.prop('component')).toBe('aside');
  });

  it('accepts a component for the total', () => {
    wrapper = mount(<HashRouter>
      <CounterBar>
        <CounterBar.Total score={2} component={Link} to="/courses?sort=name'/">All counters</CounterBar.Total>
        <CounterBar.Counters>
          <CounterBar.Counter score={1}>Counter 1</CounterBar.Counter>
          <CounterBar.Counter score={2}>Counter 2</CounterBar.Counter>
          <CounterBar.Counter score={3}>Counter 3</CounterBar.Counter>
          <CounterBar.Counter score={4}>Counter 4</CounterBar.Counter>
          <CounterBar.Counter score={5}>Counter 5</CounterBar.Counter>
        </CounterBar.Counters>
      </CounterBar>
    </HashRouter>);
    const totalWrapper = wrapper.find('Total');
    expect(totalWrapper.prop('component')).toBe(Link);
  });

  it('accepts an HTML element string for a counter', () => {
    wrapper = mount(<CounterBar>
      <CounterBar.Total score={15}>All counters</CounterBar.Total>
      <CounterBar.Counters>
        <CounterBar.Counter score={1} component="aside">Counter 1</CounterBar.Counter>
        <CounterBar.Counter score={2}>Counter 2</CounterBar.Counter>
        <CounterBar.Counter score={3}>Counter 3</CounterBar.Counter>
        <CounterBar.Counter score={4}>Counter 4</CounterBar.Counter>
        <CounterBar.Counter score={5}>Counter 5</CounterBar.Counter>
      </CounterBar.Counters>
    </CounterBar>);
    const counterWrapper = wrapper.find('Counter').first();
    expect(counterWrapper.prop('component')).toBe('aside');
  });

  it('accepts a component for a counter', () => {
    wrapper = mount(<HashRouter>
      <CounterBar>
        <CounterBar.Total score={2}>All counters</CounterBar.Total>
        <CounterBar.Counters>
          <CounterBar.Counter score={1} component={Link} to="/courses/1/">Counter 1</CounterBar.Counter>
          <CounterBar.Counter score={2}>Counter 2</CounterBar.Counter>
          <CounterBar.Counter score={3}>Counter 3</CounterBar.Counter>
          <CounterBar.Counter score={4}>Counter 4</CounterBar.Counter>
          <CounterBar.Counter score={5}>Counter 5</CounterBar.Counter>
        </CounterBar.Counters>
      </CounterBar>
    </HashRouter>);
    const counterWrapper = wrapper.find('Counter').first();
    expect(counterWrapper.prop('component')).toBe(Link);
  });

  it('matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
