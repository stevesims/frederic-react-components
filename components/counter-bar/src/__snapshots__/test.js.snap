// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CounterBar matches snapshot 1`] = `
.emotion-52 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  line-height: 1;
}

.emotion-6 {
  border: 0px;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin: 0 6px 6px 0;
  outline: none;
  padding: 0 8px 0 0;
  background: #ffffff;
}

.emotion-6:focus {
  outline: solid 4px #ffbf47;
}

.emotion-4 {
  display: inline-block;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.emotion-0 {
  background: #b10e1e;
  color: #ffffff;
  display: inline-block;
  font-family: Roboto,Arial,sans-serif;
  font-size: 15px;
  line-height: 1.25;
  min-height: 20px;
  min-width: 20px;
  padding: 1px 4px;
  text-align: center;
}

.emotion-2 {
  display: inline-block;
  font-family: Roboto,Arial,sans-serif;
  font-size: 15px;
  line-height: 1.25;
  margin-left: 10px;
}

.emotion-50 {
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-right: -6px;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.emotion-14 {
  border: 0px;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin: 0 6px 6px 0;
  outline: none;
  padding: 0 8px 0 0;
  background: #ffffff;
  color: #ffffff;
  background: #6f777b;
}

.emotion-14:focus {
  outline: solid 4px #ffbf47;
}

<HashRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <CounterBar>
      <OuterWrapper>
        <div
          className="emotion-52 emotion-53"
        >
          <Total
            component="a"
            score={2}
            scoreBackgroundColor="#b10e1e"
            scoreColor="#ffffff"
            scoreDisabledBackgroundColor="#dee0e2"
            scoreDisabledColor="#0b0c0c"
          >
            <Item
              component="a"
              score={2}
              scoreBackgroundColor="#b10e1e"
              scoreColor="#ffffff"
              scoreDisabledBackgroundColor="#dee0e2"
              scoreDisabledColor="#0b0c0c"
              wrapper={[Function]}
            >
              <TotalWrapper
                disabled={false}
                empty={0}
              >
                <a
                  className="emotion-6 emotion-7"
                  disabled={false}
                >
                  <ResultCountTitle
                    count={2}
                    countBackgroundColor="#b10e1e"
                    countColor="#ffffff"
                  >
                    <Wrapper>
                      <div
                        className="emotion-4 emotion-1"
                      >
                        <ResultCount
                          backgroundColor="#b10e1e"
                          color="#ffffff"
                        >
                          <StyledResultCount
                            backgroundColor="#b10e1e"
                            color="#ffffff"
                          >
                            <span
                              className="emotion-0 emotion-1"
                              color="#ffffff"
                            >
                              2
                            </span>
                          </StyledResultCount>
                        </ResultCount>
                        <Title>
                          <div
                            className="emotion-2 emotion-3"
                          >
                            All counters
                          </div>
                        </Title>
                      </div>
                    </Wrapper>
                  </ResultCountTitle>
                </a>
              </TotalWrapper>
            </Item>
          </Total>
          <CountersWrapper>
            <div
              className="emotion-50 emotion-51"
            >
              <Counter
                component={[Function]}
                score={1}
                scoreBackgroundColor="#b10e1e"
                scoreColor="#ffffff"
                scoreDisabledBackgroundColor="#dee0e2"
                scoreDisabledColor="#0b0c0c"
                to="/courses/1/"
              >
                <Item
                  component={[Function]}
                  score={1}
                  scoreBackgroundColor="#b10e1e"
                  scoreColor="#ffffff"
                  scoreDisabledBackgroundColor="#dee0e2"
                  scoreDisabledColor="#0b0c0c"
                  to="/courses/1/"
                  wrapper={[Function]}
                >
                  <CounterWrapper
                    disabled={false}
                    empty={0}
                    replace={false}
                    to="/courses/1/"
                  >
                    <Link
                      className="emotion-14 emotion-7"
                      disabled={false}
                      empty={0}
                      replace={false}
                      to="/courses/1/"
                    >
                      <a
                        className="emotion-14 emotion-7"
                        disabled={false}
                        empty={0}
                        href="#/courses/1/"
                        onClick={[Function]}
                      >
                        <ResultCountTitle
                          count={1}
                          countBackgroundColor="#b10e1e"
                          countColor="#ffffff"
                        >
                          <Wrapper>
                            <div
                              className="emotion-4 emotion-1"
                            >
                              <ResultCount
                                backgroundColor="#b10e1e"
                                color="#ffffff"
                              >
                                <StyledResultCount
                                  backgroundColor="#b10e1e"
                                  color="#ffffff"
                                >
                                  <span
                                    className="emotion-0 emotion-1"
                                    color="#ffffff"
                                  >
                                    1
                                  </span>
                                </StyledResultCount>
                              </ResultCount>
                              <Title>
                                <div
                                  className="emotion-2 emotion-3"
                                >
                                  Counter 1
                                </div>
                              </Title>
                            </div>
                          </Wrapper>
                        </ResultCountTitle>
                      </a>
                    </Link>
                  </CounterWrapper>
                </Item>
              </Counter>
              <Counter
                component="a"
                score={2}
                scoreBackgroundColor="#b10e1e"
                scoreColor="#ffffff"
                scoreDisabledBackgroundColor="#dee0e2"
                scoreDisabledColor="#0b0c0c"
              >
                <Item
                  component="a"
                  score={2}
                  scoreBackgroundColor="#b10e1e"
                  scoreColor="#ffffff"
                  scoreDisabledBackgroundColor="#dee0e2"
                  scoreDisabledColor="#0b0c0c"
                  wrapper={[Function]}
                >
                  <CounterWrapper
                    disabled={false}
                    empty={0}
                  >
                    <a
                      className="emotion-14 emotion-7"
                      disabled={false}
                    >
                      <ResultCountTitle
                        count={2}
                        countBackgroundColor="#b10e1e"
                        countColor="#ffffff"
                      >
                        <Wrapper>
                          <div
                            className="emotion-4 emotion-1"
                          >
                            <ResultCount
                              backgroundColor="#b10e1e"
                              color="#ffffff"
                            >
                              <StyledResultCount
                                backgroundColor="#b10e1e"
                                color="#ffffff"
                              >
                                <span
                                  className="emotion-0 emotion-1"
                                  color="#ffffff"
                                >
                                  2
                                </span>
                              </StyledResultCount>
                            </ResultCount>
                            <Title>
                              <div
                                className="emotion-2 emotion-3"
                              >
                                Counter 2
                              </div>
                            </Title>
                          </div>
                        </Wrapper>
                      </ResultCountTitle>
                    </a>
                  </CounterWrapper>
                </Item>
              </Counter>
              <Counter
                component="a"
                score={3}
                scoreBackgroundColor="#b10e1e"
                scoreColor="#ffffff"
                scoreDisabledBackgroundColor="#dee0e2"
                scoreDisabledColor="#0b0c0c"
              >
                <Item
                  component="a"
                  score={3}
                  scoreBackgroundColor="#b10e1e"
                  scoreColor="#ffffff"
                  scoreDisabledBackgroundColor="#dee0e2"
                  scoreDisabledColor="#0b0c0c"
                  wrapper={[Function]}
                >
                  <CounterWrapper
                    disabled={false}
                    empty={0}
                  >
                    <a
                      className="emotion-14 emotion-7"
                      disabled={false}
                    >
                      <ResultCountTitle
                        count={3}
                        countBackgroundColor="#b10e1e"
                        countColor="#ffffff"
                      >
                        <Wrapper>
                          <div
                            className="emotion-4 emotion-1"
                          >
                            <ResultCount
                              backgroundColor="#b10e1e"
                              color="#ffffff"
                            >
                              <StyledResultCount
                                backgroundColor="#b10e1e"
                                color="#ffffff"
                              >
                                <span
                                  className="emotion-0 emotion-1"
                                  color="#ffffff"
                                >
                                  3
                                </span>
                              </StyledResultCount>
                            </ResultCount>
                            <Title>
                              <div
                                className="emotion-2 emotion-3"
                              >
                                Counter 3
                              </div>
                            </Title>
                          </div>
                        </Wrapper>
                      </ResultCountTitle>
                    </a>
                  </CounterWrapper>
                </Item>
              </Counter>
              <Counter
                component="a"
                score={4}
                scoreBackgroundColor="#b10e1e"
                scoreColor="#ffffff"
                scoreDisabledBackgroundColor="#dee0e2"
                scoreDisabledColor="#0b0c0c"
              >
                <Item
                  component="a"
                  score={4}
                  scoreBackgroundColor="#b10e1e"
                  scoreColor="#ffffff"
                  scoreDisabledBackgroundColor="#dee0e2"
                  scoreDisabledColor="#0b0c0c"
                  wrapper={[Function]}
                >
                  <CounterWrapper
                    disabled={false}
                    empty={0}
                  >
                    <a
                      className="emotion-14 emotion-7"
                      disabled={false}
                    >
                      <ResultCountTitle
                        count={4}
                        countBackgroundColor="#b10e1e"
                        countColor="#ffffff"
                      >
                        <Wrapper>
                          <div
                            className="emotion-4 emotion-1"
                          >
                            <ResultCount
                              backgroundColor="#b10e1e"
                              color="#ffffff"
                            >
                              <StyledResultCount
                                backgroundColor="#b10e1e"
                                color="#ffffff"
                              >
                                <span
                                  className="emotion-0 emotion-1"
                                  color="#ffffff"
                                >
                                  4
                                </span>
                              </StyledResultCount>
                            </ResultCount>
                            <Title>
                              <div
                                className="emotion-2 emotion-3"
                              >
                                Counter 4
                              </div>
                            </Title>
                          </div>
                        </Wrapper>
                      </ResultCountTitle>
                    </a>
                  </CounterWrapper>
                </Item>
              </Counter>
              <Counter
                component="a"
                score={5}
                scoreBackgroundColor="#b10e1e"
                scoreColor="#ffffff"
                scoreDisabledBackgroundColor="#dee0e2"
                scoreDisabledColor="#0b0c0c"
              >
                <Item
                  component="a"
                  score={5}
                  scoreBackgroundColor="#b10e1e"
                  scoreColor="#ffffff"
                  scoreDisabledBackgroundColor="#dee0e2"
                  scoreDisabledColor="#0b0c0c"
                  wrapper={[Function]}
                >
                  <CounterWrapper
                    disabled={false}
                    empty={0}
                  >
                    <a
                      className="emotion-14 emotion-7"
                      disabled={false}
                    >
                      <ResultCountTitle
                        count={5}
                        countBackgroundColor="#b10e1e"
                        countColor="#ffffff"
                      >
                        <Wrapper>
                          <div
                            className="emotion-4 emotion-1"
                          >
                            <ResultCount
                              backgroundColor="#b10e1e"
                              color="#ffffff"
                            >
                              <StyledResultCount
                                backgroundColor="#b10e1e"
                                color="#ffffff"
                              >
                                <span
                                  className="emotion-0 emotion-1"
                                  color="#ffffff"
                                >
                                  5
                                </span>
                              </StyledResultCount>
                            </ResultCount>
                            <Title>
                              <div
                                className="emotion-2 emotion-3"
                              >
                                Counter 5
                              </div>
                            </Title>
                          </div>
                        </Wrapper>
                      </ResultCountTitle>
                    </a>
                  </CounterWrapper>
                </Item>
              </Counter>
            </div>
          </CountersWrapper>
        </div>
      </OuterWrapper>
    </CounterBar>
  </Router>
</HashRouter>
`;
